### PR TITLE
Hide implementation of Generation

### PR DESCRIPTION
**Pull request by futile**
Instead, expose `is_alive()` and `is_dead()` methods which have a clear
name, and refactor existing code to use these.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -23,13 +23,24 @@ pub use world::{Component, World, FetchArg,
...
+pub struct Generation(i32);
+
+impl Generation {
+    /// Returns `true` if entities of this generation are alive
+    pub fn is_alive(&self) -> bool {
+        self.0 > 0
+    }
+
+    /// Returns `true` if entities of this generation are dead
+    pub fn is_dead(&self) -> bool {
```
Comment (kvark) [inline src/lib.rs]: I'd rather not have a separate method here, `!x.is_alive()` is clear enough

File: src/world.rs [hunk 2]
```diff
@@ -59,14 +59,14 @@ struct Appendix {
 }
 
 fn find_next(gens: &[Generation], lowest_free_index: usize) -> Entity {
-    if let Some((id, gen)) = gens.iter().enumerate().skip(lowest_free_index).find(|&(_, g)| *g <= 0) {
-        return Entity(id as Index, 1 - gen);
+    if let Some((id, gen)) = gens.iter().enumerate().skip(lowest_free_index).find(|&(_, g)| g.is_dead()) {
+        return Entity(id as Index, Generation(1 - gen.0));
```
Comment (kvark) [inline src/world.rs]: we better have a method here, like `gen.die()`

File: src/world.rs [hunk 3]
```diff
@@ -59,14 +59,14 @@ struct Appendix {
...
     }
 
     if lowest_free_index > gens.len() {
-        Entity(lowest_free_index as Index, 1)
+        Entity(lowest_free_index as Index, Generation(1))
```
Comment (kvark) [inline src/world.rs]: Please refactor this to only return the index id from the condition, and then return the `Entity` in a separate line

File: src/world.rs [hunk 4]
```diff
@@ -219,12 +219,12 @@ impl World {
         }
         let mut gens = self.generations.write().unwrap();
         let mut gen = &mut gens[entity.get_id() as usize];
-        assert!(*gen > 0);
+        assert!(gen.is_alive());
         let mut app = self.appendix.write().unwrap();
         if entity.get_id() < app.next.get_id() {
-            app.next = Entity(entity.0, *gen+1);
+            app.next = Entity(entity.0, Generation(gen.0 + 1));
```
Comment (kvark) [inline src/world.rs]: This was rather a hack - to do `gen.0+1` before the `gen` is inverted. Instead, we should do:
```rust
gen = gen.die();
...
app.next = Entity(..., gen.raise());
```

File: src/world.rs [hunk 5]
```diff
@@ -253,19 +253,19 @@ impl World {
...
+                assert_eq!(ent.get_gen().0, 1 - gens[ent.get_id()].0);
                 gens[ent.get_id()] = ent.get_gen();
             }
             let mut next = app.next;
             for ent in app.sub_queue.drain(..) {
                 assert_eq!(ent.get_gen(), gens[ent.get_id()]);
                 temp_list.push(ent);
                 if ent.get_id() < next.get_id() {
-                    next = Entity(ent.0, ent.1 + 1);
+                    next = Entity(ent.0, Generation(ent.get_gen().0 + 1));
```
Comment (kvark) [inline src/world.rs]: same here (as my previous comment)

File: src/world.rs (hunk 4 above)
Comment (futile) [inline src/world.rs]: `gen.raise()` should basically return `Generation(gen.0 + 1)`?

**Comment by kvark** [comment]
Thanks a bunch!

**Pull request merged**